File: 518_coin_change_ii/coin_change_ii.go
package main

import "fmt"

type Solution struct{}

func (s *Solution) change_change_ii(amount int, coins []int) int {
	dp := make([]int, amount+1)
	dp[0] = 1

	for _, coin := range coins {
		for x := coin; x <= amount; x++ {
			dp[x] += dp[x-coin]
		}
	}
	return dp[amount]
}

func main() {
	s := Solution{}
	fmt.Println(s.change_change_ii(5, []int{1, 2, 5}))
}

File: 518_coin_change_ii/coin_change_ii.py
from typing import List

class Solution:
    def coin_change_ii(self, amount: int, coins: List[int]) -> int:
        dp = [0] * (amount + 1)
        dp[0] = 1

        for coin in coins:
            for x in range(coin, amount + 1):
                dp[x] += dp[x - coin]

        return dp[amount]


if __name__ == "__main__":
    solution = Solution()
    print(solution.coin_change_ii(5, [1, 2, 5]))
